carbon/feed.py

```diff
@@ -0,0 +1,341 @@
+from abc import ABC, abstractmethod
+from collections.abc import Generator
+from contextlib import closing
+from datetime import datetime
+from typing import IO, Any, ClassVar
+
+from lxml import etree as ET
+from sqlalchemy import func, select
+from sqlalchemy.sql.selectable import Select
+
+from carbon.database import DatabaseEngine, aa_articles, dlcs, orcids, persons
+from carbon.helpers import (
+    get_group_name,
+    get_hire_date_string,
+    get_initials,
+)
+
+
+class BaseXmlFeed(ABC):
+    """Base XML feed class.
+
+    This is the abstract class for creating XML feeds. The following class attributes
+    are unique per subclass of carbon.feed.BaseXmlFeed:
+
+        root_element_name: The 'tag' assigned to the root Element.
+        query: The select statmenet submitted to the Data Warehouse to retrieve records.
+
+    Attributes:
+        engine: A configured carbon.database.DatabaseEngine that can connect to the
+            Data Warehouse.
+        output_file: A file-like object (stream) into which normalized XML strings
+            strings are written.
+
+    """
+
+    root_element_name: str | ET.QName = ""
+    query: Select = select()
+    record_count: int = 0
+
+    def __init__(self, engine: DatabaseEngine, output_file: IO):
+        self.engine = engine
+        self.output_file = output_file
+
+    @property
+    def records(self) -> Generator[dict[str, Any], Any, None]:
+        """Create a generator of 'people' or 'article' records from the Data Warehouse.
+
+        Yields:
+            Generator[dict[str, Any], Any, None]: Records that
+                match the query submitted to the Data Warehouse.
+        """
+        with closing(self.engine().connect()) as connection:
+            result = connection.execute(self.query)
+            for row in result:
+                self.record_count += 1
+                yield dict(zip(result.keys(), row, strict=True))
+
+    @abstractmethod
+    def _add_element(self, record: dict[str, Any]) -> None | ET._Element:  # noqa: SLF001
```

Comment:
Is there a better way to do this type hinting than with a private member? I'm wary of introducing so many linter overrides, does `ET.Element` not work for some reason?

Reply:
When I use `ET.Element`, `mypy` throws the following error: 

> carbon/feed.py:59: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]


1. Came across these StackOverflow threads discussing the issue: [this](https://stackoverflow.com/questions/72226485/mypy-function-lxml-etree-elementtree-is-not-valid-as-a-type-but-why), [this](https://stackoverflow.com/questions/76321801/type-hints-for-lxml-etree-element), and [this](https://stackoverflow.com/questions/70310417/correct-way-to-type-hint-protected-private-classes-in-python).

3. Here is the [class definition](https://github.com/lxml/lxml/blob/53c5a224a4e6f8209a063ebc003cf296c5844b43/src/lxml/etree.pyx#L1871-L1872) for `lxml.etree._Element`.

My takeaway: It seems like the only option available with `lxml-stubs` is using these "private" Cython classes. 🤔 I would opt that we ignore this "style" error (as we are currently doing) instead of having `mypy` ignore it completely. Let me know what you think!

Reply:
With further context, that seems fine